### PR TITLE
Improve performance, for real this time.

### PR DESCRIPTION
**Pull request by kosayoda**
In particular, the PR decreases CPU usage of the release binary from 200% on my machine to 10%.

`main`:
image

`performance`:
image

10% is still high, but this is yet another step in the right direction.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -22,8 +22,8 @@ async fn main() -> Result<()> {
 
         let app = App::new(config.terminal.maximum_messages as usize);
 
-        let (twitch_tx, terminal_rx) = mpsc::channel(1);
-        let (terminal_tx, twitch_rx) = mpsc::channel(1);
+        let (twitch_tx, terminal_rx) = mpsc::channel(100);
+        let (terminal_tx, twitch_rx) = mpsc::channel(100);
```
Comment (Xithrius) [inline src/main.rs:26]: Why are we increasing this to 100?

File: src/twitch.rs [hunk 2]
```diff
@@ -27,44 +24,48 @@ pub async fn twitch_irc(config: &CompleteConfig, tx: Sender<Data>, mut rx: Recei
...
-                        true,
-                    ))
-                    .await
-                    .unwrap();
-                }
-                _ => (),
-            }
-        }
+        tokio::select! {
+            biased;
```
Comment (Xithrius) [inline src/twitch.rs:28]: What does `biased;` do?

File: src/twitch.rs [hunk 3]
```diff
@@ -27,44 +24,48 @@ pub async fn twitch_irc(config: &CompleteConfig, tx: Sender<Data>, mut rx: Recei
...
 
-        if let Some(Some(message)) = unconstrained(rx.recv()).now_or_never() {
-            client
+            Some(message) = rx.recv() => {
```
Comment (Xithrius) [inline src/twitch.rs:30]: Making sure we get a message through `recv` instead of `try_recv`, yeah?

File: src/twitch.rs [hunk 4]
```diff
@@ -27,44 +24,48 @@ pub async fn twitch_irc(config: &CompleteConfig, tx: Sender<Data>, mut rx: Recei
...
+                        ))
+                        .await
+                        .unwrap();
+                    }
+                    Command::NOTICE(ref _target, ref msg) => {
+                        tx.send(Data::new(
+                            Local::now()
+                                .format(config.frontend.date_format.as_str())
+                                .to_string(),
+                            "twitch".to_string(),
```
Comment (Xithrius) [inline src/twitch.rs:59]:
```suggestion
                            "Twitch".to_string(),
```

File: src/utils/event.rs [hunk 5]
```diff
@@ -56,25 +53,10 @@ impl Events {
...
-                    }
-                    tokio::time::sleep(config.tick_rate).await;
-                }
-            })
-        };
-        Events {
-            rx,
-            input_handle,
-            tick_handle,
-        }
```
Comment (Xithrius) [inline src/utils/event.rs:74]: I'm wondering why I never removed this in the first place. Thanks for obliterating it!

**Review by Xithrius — CHANGES_REQUESTED** [review]
Tis the good stuff.

File: src/main.rs (hunk 1 above)
Comment (kosayoda) [reply src/main.rs:26]: 100 is arbitrary, anything above 1 is sensible. A channel buffer of 1 will make `.send`s sleep until the message is removed by the receiver. Thus, it doesn't make sense to only have one message in queue.

File: src/twitch.rs (hunk 2 above)
Comment (kosayoda) [reply src/twitch.rs:28]: It lets me [https://docs.rs/tokio/1.12.0/tokio/macro.select.html#fairness](control polling order). This is partly for ease of mind so we always poll for a new message to send to Twitch before polling for new messages from Twitch.

File: src/main.rs (hunk 1 above)
Comment (Xithrius) [reply src/main.rs:26]: I see. Thank you for the confirmation.

File: src/twitch.rs (hunk 3 above)
Comment (kosayoda) [reply src/twitch.rs:30]: `select!` waits on multiple branches concurrently, so yes this won't block the twitch handler.

File: src/twitch.rs (hunk 2 above)
Comment (Xithrius) [reply src/twitch.rs:28]: Very well.

File: src/twitch.rs (hunk 3 above)
Comment (Xithrius) [reply src/twitch.rs:30]: Very cool.

**Review by Xithrius — APPROVED** [review]
Oh snap, apparently it really does burn my CPU. I was checking the incorrect place.

Current main branch:
image

This branch:
image

**Pull request merged**